src/EditorFeatures/Core/Shared/Extensions/OperationContextExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
+{
+    internal static class OperationContextExtensions
+    {
+        /// <summary>
+        /// Wait until given document is fully loaded and then add scope
+        /// </summary>
+        public static IUIThreadOperationScope WaitForDocumentAndAddScope(
+            this IUIThreadOperationContext context, Document document, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, document.Project.Solution, document.Name, allowCancellation, description);
+        }
+
+        public static IUIThreadOperationScope WaitForProjectAndAddScope(
+            this IUIThreadOperationContext context, Project project, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, project.Solution, project.Name, allowCancellation, description);
+        }
+
+        private static IUIThreadOperationScope WaitAndAddScope(
+            this IUIThreadOperationContext context, Solution solution, string name, bool allowCancellation, string description)
+        {
+            IUIThreadOperationScope scope = null;
+
+            try
+            {
+                // partial mode is always cancellable
+                var title = string.Format(EditorFeaturesResources.Operation_is_not_ready_for_0_yet_see_task_center_for_more_detail, name);
+                scope = context.AddScope(allowCancellation: true, title);
+
+                var service = solution.Workspace.Services.GetService<ISolutionStatusService>();
+                if (service != null)
+                {
+                    // UIThreadOperation always called from UI thread.
+                    service.WaitForAsync(solution, scope.Context.UserCancellationToken).Wait(context.UserCancellationToken);
+                }
+
+                // now set description and cancellation as asked
+                scope.AllowCancellation = allowCancellation;
+                scope.Description = description;
+
+                return scope;
+            }
+            catch (OperationCanceledException)
+            {
+                // cancellation is exepcted. dispose scopes and rethrow exceptions
+                scope?.Dispose();
+
+                throw;
```

Comment:
this is not safe.  You are breaking TPL invariants.  Specifically, because you made an inner cancellation token, you are allowing an OCE to leak out based on that CT.  However, callers higher up in the stack may only allow OCE's assocaited with their own cancellation token.  If this leaks out, it will cause an AggregateException and will move the task into the faulted state.

Reply:
this is rethrow, no idea what you are talking about, it should throw same exception as it got. and this is not under TPL.

to double check, I just did this

```csharp
static void Main(string[] args)
        {
            var source = new CancellationTokenSource();

            try
            {
                source.Cancel();
                try
                {
                    source.Token.ThrowIfCancellationRequested();

                }
                catch (Exception ex)
                {
                    throw;
                }
            }
            catch (OperationCanceledException ce)
            {
                if (ce.CancellationToken  == source.Token)
                {

                }

            }
        }
```

and I get same cancellation token as the one thrown.

Reply:
You may have a scope taht does not allow cancellation.  You are adding a scope that adds cancellation, and then allowing that OCE to pass outwards.  That may not be in a context that supports cancellation.  That would break invariants of the caller as they may not be expecting this.  

i.e. if i call "Foo(a, b)" and i'm not passing a cancellation token, i hsould not get some randome cancellation exception passing back out.

Reply:
>to double check, I just did this

Your example is invalid.  Consider instead:

```c# 
static void Main(string[] args)
{
    Foo();
    Console.WriteLine("I should always get here");
}

static void Foo()
{
            var source = new CancellationTokenSource();
            source.Cancel();
            source.Token.ThrowIfCancellationRequested();
}
```

The person calling Foo has no expectation that cancellation is ever possible.  You've now violated their invariants.  Consider, for example, if it was even this:


```c#
        public static Task Main()
        {
            return BarAsync(default);
        }

        static Task BarAsync(CancellationToken ct)
        {
            return Task.Run(() => Foo(), ct).ContinueWith(_ => Console.WriteLine("I should always get here"));
        }

        static void Foo()
        {
            var source = new CancellationTokenSource();
            source.Cancel();
            source.Token.ThrowIfCancellationRequested();
        }
```

You will end up with tasks in the faulted state.  Because you are throwing cancellation tokens that were unexpected.

Reply:
we already have a lot of code that accepts cancellation as field and throw cancelaltion from the field. no way when user call method from the instance passed down from someone know that it will throw or not unless one know the code.

if you want to add more documentation, sure. I can add. otherwise, I don't agree.

Reply:
also, what you are saying is not true, since it passes down from context which has cancellation token in it. we have already code all around where it pass down the context and expect cancellation to thrown. just see Execute, it just pass down context and we 100% expect it to throw cancellation.

Reply:
> we have already code all around where it pass down the context and expect cancellation to thrown

I don't believe that we have any code that *disables* cancellation, which then expects that someone else down the line may *enable* cancellation.  

Furthermore, i don't see why you need to change *anything*.  Either the calling function is one that accepts cancellation, in which case *you don't need to do anything*.  Or, they don't accept cancellation, in which case *you don't need to do anything*.  What i don't understand is: why are you disregarding the wishes of the caller?

If the caller said "i'm not ok with cancellation", you are now overriding it.  If you then throw cancellation, you will pass a cancellation token to hte caller which already said it wasn't ok with that.

As the person enabling cancellation, you should not leak that cancellation token outwards.

Reply:
waiting for a partial load is always cancellable. if we ever have a case where waiting for partial mode is not cancellable, we will address that at the time.

Reply:
See my suggestion on how you can acocmplish this without breaking the invariants of callers.